From 2aa57712bf47bdc81ba7cee573d1d9f7ec24ab09 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:48:53 -0400
Subject: [tools] perf/script: Print callchains and symbols if they exist

Message-id: <1408953210-26343-46-git-send-email-jolsa@redhat.com>
Patchwork-id: 88755
O-Subject: [PATCH RHEL7.1 BZ1133083 045/322] perf script: Print callchains and symbols if they exist
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 80b8b496ec6edaff01f9ab74dbe8a517cd718de8
Author: David Ahern <dsahern@gmail.com>
Date: Tue Nov 19 21:07:37 2013 -0700

description
===========
The intent of perf-script is to dump the events and information in the
file. H/W, S/W and raw events all dump callchains if they are present;
might as well make that the default for tracepoints too.

v2: Only add options for sym, dso and ip if callchains are present
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-script.c b/tools/perf/builtin-script.c
index b392770..9f3ba44 100644
--- a/tools/perf/builtin-script.c
+++ b/tools/perf/builtin-script.c
@@ -280,6 +280,30 @@ static int perf_session__check_output_opt(struct perf_session *session)
   set_print_ip_opts(&evsel->attr);
  }
 
+ /*
+  * set default for tracepoints to print symbols only
+  * if callchains are present
+  */
+ if (symbol_conf.use_callchain &&
+     !output[PERF_TYPE_TRACEPOINT].user_set) {
+  struct perf_event_attr *attr;
+
+  j = PERF_TYPE_TRACEPOINT;
+  evsel = perf_session__find_first_evtype(session, j);
+  if (evsel == NULL)
+   goto out;
+
+  attr = &evsel->attr;
+
+  if (attr->sample_type & PERF_SAMPLE_CALLCHAIN) {
+   output[j].fields |= PERF_OUTPUT_IP;
+   output[j].fields |= PERF_OUTPUT_SYM;
+   output[j].fields |= PERF_OUTPUT_DSO;
+   set_print_ip_opts(attr);
+  }
+ }
+
+out:
  return 0;
 }
 
-- 
1.7.1